drivers/spi/spi_sc18is606.c

```diff
@@ -0,0 +1,259 @@
+
+		/*If the read failed set ret to EAGAIN*/
+		ret = ret == -1 ? -EAGAIN : ret;
+
+		if (ret < 0) {
+			LOG_ERR("Failed to read data (%d)", ret);
+		}
+	}
+
+	k_mutex_unlock(&data->bridge_lock);
+
+	return 0;
+}
+
+static int sc18is606_spi_configure(const struct device *dev, const struct spi_config *config)
+{
+	struct nxp_sc18is606_data *data = dev->data;
+
+	if (config->operation & SPI_OP_MODE_SLAVE) {
+		LOG_ERR("SC18IS606 does not support Slave mode");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & (SPI_LINES_DUAL | SPI_LINES_QUAD | SPI_LINES_OCTAL)) {
+		LOG_ERR("Unsupported line configuration");
+		return -ENOTSUP;
+	}
+
+	const int bits = SPI_WORD_SIZE_GET(config->operation);
+
+	if (bits > 8) {
+		LOG_ERR("Word sizes > 8 bits not supported");
+		return -ENOTSUP;
+	}
+
+	/* Build SC18IS606  configuration byte*/
+	uint8_t cfg_byte = 0;
+
+	cfg_byte |= ((config->operation & SPI_TRANSFER_LSB) >> 4) << 5;
+
+	uint8_t mode = (SPI_MODE_GET(config->operation) >> 1) & 0x3;
+
+	cfg_byte |= (mode << 2);
+
+	uint8_t freq;
+	if (config->frequency >= 1800000) {
+		freq = 0b00;
+	} else if (config->frequency >= 400000) {
+		freq = 0b01;
+	} else if (config->frequency >= 100000) {
+		freq = 0b10;
+	} else {
+		freq = 0b11;
+	}
+
+	cfg_byte |= freq;
+
+	data->ctx.config = config;
+
+	uint8_t buffer[2] = {SC18IS606_CONFIG_SPI, cfg_byte};
+
+	cfg_byte = nxp_sc18is606_transfer(dev, buffer, sizeof(buffer), NULL, 0);
+
+	return 0;
+}
+
+static int sc18is606_spi_transceive(const struct device *dev, const struct spi_config *spi_cfg,
+				    const struct spi_buf_set *tx_buffer_set,
+				    const struct spi_buf_set *rx_buffer_set)
+{
+	int ret = 1;
+	int reconfig = 1;
+
+	reconfig = sc18is606_spi_configure(dev, spi_cfg);
+	if (reconfig < 0) {
+		return reconfig;
+	}
+
+	if (!tx_buffer_set && !rx_buffer_set) {
+		LOG_ERR("SC18IS606 at least one buffer_set should be set \n");
+		return -EINVAL;
+	}
+
+	/* CS line to be Used */
+	uint8_t ss_idx = spi_cfg->slave;
+	if (ss_idx > 2) {
+		LOG_ERR("SC18IS606: Invalid SS Index (%u) must be 0-2", ss_idx);
+		return -EINVAL;
+	}
+
+	uint8_t function_id = (1 << ss_idx) & 0x07;
+
+	if (tx_buffer_set && tx_buffer_set->buffers && tx_buffer_set->count > 0) {
+		for (size_t i = 0; i < tx_buffer_set->count; i++) {
+			const struct spi_buf *tx_buf = &tx_buffer_set->buffers[i];
+			const size_t len = tx_buf->len;
+
+			uint8_t sc18_buf[1 + len];
+			sc18_buf[0] = function_id;
+			memcpy(&sc18_buf[1], tx_buf->buf, len);
+
+			ret = nxp_sc18is606_transfer(dev, sc18_buf, sizeof(sc18_buf), NULL, 0);
+		}
+	}
+
+	if (rx_buffer_set && rx_buffer_set->buffers && rx_buffer_set->count > 0) {
+		for (size_t i = 0; i < rx_buffer_set->count; i++) {
+			/* Function ID first to select the device */
+			uint8_t cmd_buf[1] = {function_id};
+			const struct spi_buf *rx_buf = &rx_buffer_set->buffers[i];
+			ret = nxp_sc18is606_transfer(dev, cmd_buf, sizeof(cmd_buf), rx_buf->buf,
```

Comment:
It seems like this assignment is shadowing the previous `ret = ` on the block above?

What if the TX fails but the RX succeeds?

Maybe/likey I missed something.

Reply:
A read needs to start with a tx for this device, I may also have missed something.

Reply:
You could eventually switch to the error handling idiom `ret = nxp_sc18is606_transfer(...); if (ret < 0) { LOG_ERR(...); return ret; }` here and above.

Then for instance if the 1st buffer has an error and the 2nd buffer succeeds, currently it returns success?

Reply:
That's in  now